File: HW/HW 11/FloodFill.py
class FloodFill:
    def isValidChecker(self, imageMatrix, row, col, color):
        return (imageMatrix and len(imageMatrix) > 0 and len(imageMatrix[0]) > 0 and imageMatrix[row][col] != color)

    def checkConstraints(self, image, sr, sc, color):
        m, n = len(image), len(image[0])
        counter = sum(1 for row in image for pixel in row if pixel < 0)
        return (1 <= m <= 50 and 0 <= sr < m and 0 <= sc < n and counter == 0)

    def floodFill(self, image, sr, sc, newColor):
        if not self.isValidChecker(image, sr, sc, newColor):
            return image
        if not  self.checkConstraints(image, sr, sc, newColor):
            raise ValueError("Input does not meet the constraints")

        m, n = len(image), len(image[0])
        originalColor = image[sr][sc]
        floodFillQueue = [(sr, sc)]

        while floodFillQueue:
            rowNum, colNum = floodFillQueue.pop(0)
            image[rowNum][colNum] = newColor

            if rowNum - 1 >= 0 and image[rowNum - 1][colNum] == originalColor:
                floodFillQueue.append((rowNum - 1, colNum))
            if rowNum + 1 < m and image[rowNum + 1][colNum] == originalColor:
                floodFillQueue.append((rowNum + 1, colNum))
            if colNum - 1 >= 0 and image[rowNum][colNum - 1] == originalColor:
                floodFillQueue.append((rowNum, colNum - 1))
            if colNum + 1 < n and image[rowNum][colNum + 1] == originalColor:
                floodFillQueue.append((rowNum, colNum + 1))

        return image

    def display(self, image):
        try:
            for row in image:
                print(" ".join(str(pixel) for pixel in row))
        except TypeError:
            print("The matrix is null")

if __name__ == "__main__":
    obColorFill = FloodFill()

    # Test the function with the provided test cases
    image1 = [[1, 1, 1], [1, 1, 0], [1, 0, 1]]
    sr1, sc1, color1 = 1, 1, 2
    print("Input 1:")
    obColorFill.display(image1)
    result1 = obColorFill.floodFill(image1, sr1, sc1, color1)
    print("Output 1:")
    obColorFill.display(result1)

    image2 = [[0, 0, 0], [0, 0, 0]]
    sr2, sc2, color2 = 0, 0, 0
    print("Input 2:")
    obColorFill.display(image2)
    result2 = obColorFill.floodFill(image2, sr2, sc2, color2)
    print("Output 2:")
    obColorFill.display(result2)

File: HW/HW 11/FloodFill.java
import java.util.*;

public class FloodFill {
    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        if (!isValdChecker(image, sr, sc, color)) {
            return image;
        }
        if (!checkConstraints(image, sr, sc, color)) {
            throw new IllegalArgumentException("Input does not meet the constraints");
        }

        int m = image.length;
        int n = image[0].length;
        int rowNum, colNum;
        int originalColor = image[sr][sc];
        LinkedList<int[]> floodFillQueue = new LinkedList<>();
        floodFillQueue.add(new int[] { sr, sc });

        while (floodFillQueue.size() > 0) {
            int[] queueElement = floodFillQueue.removeFirst();
            rowNum = queueElement[0];
            colNum = queueElement[1];
            image[rowNum][colNum] = color;

            if (rowNum - 1 >= 0 && image[rowNum - 1][colNum] == originalColor) {
                floodFillQueue.add(new int[] { rowNum - 1, colNum });
            }
            if (rowNum + 1 < m && image[rowNum + 1][colNum] == originalColor) {
                floodFillQueue.add(new int[] { rowNum + 1, colNum });
            }
            if (colNum - 1 >= 0 && image[rowNum][colNum - 1] == originalColor) {
                floodFillQueue.add(new int[] { rowNum, colNum - 1 });
            }
            if (colNum + 1 < n && image[rowNum][colNum + 1] == originalColor) {
                floodFillQueue.add(new int[] { rowNum, colNum + 1 });
            }
        }

        return image;
    }

    public boolean isValdChecker(int[][] imageMatrix, int row, int col, int color) {
        return (imageMatrix != null && imageMatrix.length != 0 && imageMatrix[0].length != 0
                && imageMatrix[row][col] != color);
    }

    public boolean checkConstraints(int[][] image, int sr, int sc, int color) {
        int m = image.length;
        int n = image[0].length;
        int counter = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (image[i][j] < 0)
                    counter++;
            }
        }
        return ((1 <= m) && (n <= 50) && (color < Math.pow(2, 16)) &&
                (0 <= sr) && (sr < m) && (0 <= sc) && (sc < n) && (counter == 0));
    }

    public void display(int[][] image) {
        try {
            for (int i = 0; i < image.length; i++) {
                for (int j = 0; j < image[0].length; j++) {
                    System.out.print(image[i][j] + " ");
                }
                System.out.println();
            }
        } catch (NullPointerException e) {
            System.out.println("The matrix is null");
        }
    }

    public static void main(String[] args) {
        FloodFill obColorFill = new FloodFill();

        // Test the function with the provided test cases
        int[][] image1 = { { 1, 1, 1 }, { 1, 1, 0 }, { 1, 0, 1 } };
        int sr1 = 1, sc1 = 1, color1 = 2;
        System.out.println("Input 1:");
        obColorFill.display(image1);
        int[][] result1 = obColorFill.floodFill(image1, sr1, sc1, color1);
        System.out.println("Output 1:");
        obColorFill.display(result1);

        int[][] image2 = { { 0, 0, 0 }, { 0, 0, 0 } };
        int sr2 = 0, sc2 = 0, color2 = 0;
        System.out.println("Input 2:");
        obColorFill.display(image2);
        int[][] result2 = obColorFill.floodFill(image2, sr2, sc2, color2);
        System.out.println("Output 2:");
        obColorFill.display(result2);

    }
}

